src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -94,9 +96,12 @@ private bool EnqueueCore(TElement value)
                 waiter = _waiters.Dequeue();
             }
 
-            // Invoke SetResult on a separate task, as this invocation could cause the underlying task to executing,
-            // which could be a long running operation that can potentially cause a deadlock if executed on the current thread.
-            Task.Run(() => waiter.SetResult(value));
+            Debug.Assert(waiter.Task.CreationOptions.HasFlag(TaskCreationOptions.RunContinuationsAsynchronously));
```

Comment:
Contract instead of Debug to make a stronger check?

Reply:
These values are not produced by code outside this class. Switching to `Contract` has the undesirable side effect of no longer being able to use `HasFlag`, so it seemed more concise to have it in this form.

Reply:
> Switching to Contract has the undesirable side effect of no longer being able to use HasFla

I'm not sure what that means, can you clarify?

Reply:
`HasFlag` boxes enum values on .NET Framework, which is more overhead that we want at least for the `EnqueueCore` case.